### PR TITLE
#399 add refund endpoint

### PR DESCRIPTION
**Pull request by lojzatran**
In this PR, I added a refund endpoint to the Adyen integration.

TODO:
- [x] documentation

**What won't be included in this PR**
Updating the notification module - this will be done in a different PR. It means that this feature is not fully functional yet.

Part of the issue: https://github.com/commercetools/commercetools-adyen-integration/issues/399

After the endpoint is called with the following payment, refunds were made on Adyen:
```json
{
  "key": "14",
  "amountPlanned": {
    "type": "centPrecision",
    "currencyCode": "EUR",
    "centAmount": 1000,
    "fractionDigits": 2
  },
  "paymentMethodInfo": {
    "paymentInterface": "ctp-adyen-integration"
  },
  "transactions": [
    {
      "id": "3d4f784b-0e7e-489b-a52d-8b83515c4be6",
      "type": "Authorization",
      "amount": {
        "type": "centPrecision",
        "currencyCode": "EUR",
        "centAmount": 1000,
        "fractionDigits": 2
      },
      "interactionId": "883605783405120H",
      "state": "Success"
    },
    {
      "id": "19ed13c8-733e-479b-a4f9-c22e6bed7ea3",
      "type": "Charge",
      "amount": {
        "type": "centPrecision",
        "currencyCode": "EUR",
        "centAmount": 1000,
        "fractionDigits": 2
      },
      "interactionId": "883605782632488F",
      "state": "Success"
    },
    {
      "type": "Refund",
      "amount": {
        "type": "centPrecision",
        "currencyCode": "EUR",
        "centAmount": 200,
        "fractionDigits": 2
      },
      "state": "Initial"
    },
    {
      "type": "Refund",
      "amount": {
        "type": "centPrecision",
        "currencyCode": "EUR",
        "centAmount": 100,
        "fractionDigits": 2
      },
      "state": "Initial"
    },
    {
      "type": "Refund",
      "amount": {
        "type": "centPrecision",
        "currencyCode": "EUR",
        "centAmount": 300,
        "fractionDigits": 2
      },
      "state": "Initial"
    }
  ],
  "interfaceInteractions": [
  ]
}
```

Screenshot 2020-11-19 at 17 22 18

### CODE REVIEW

File: extension/docs/Refund.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
+**Please see [Integration Guide](WebComponentsIntegrationGuide.md) first before continuing with this document.**
+
+## Refund
+If you want to return the funds to your shopper, for example if they returned an item, you need to make a Refund request.
+
+### Make an API call to refund a payment:
```
Comment (lojzatran) [inline extension/docs/Refund.md:6]:
```suggestion
### Make an API call to refund a payment
```

File: extension/docs/Refund.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+#### Prerequisites
+It is required that the payment has one `Charge` transaction with state `Success`. From this transaction, the `interactionId` field is being used as `pspReference` for the Adyen refund request.
```
Comment (lojzatran) [inline extension/docs/Refund.md:9]:
```suggestion
It is required that the payment has one `Charge` transaction with state `Success`. From this transaction, the `interactionId` field is being used as `originalReference` for the Adyen refund request.
```

File: extension/docs/Refund.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    "amount": {
+      "currencyCode": "EUR",
+      "centAmount": 500
+    },
+    "state": "Initial"
+  }
+}
+```
+
+Extension module will update the transaction with `Pending` transaction state and `interactionId` field with value `pspReference` from the Adyen response:
```
Comment (lojzatran) [inline extension/docs/Refund.md:28]:
```suggestion
Extension module will update the transaction with `Pending` transaction state and `interactionId` field with the value `pspReference` from the Adyen response:
```

File: extension/docs/Refund.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        "response": "{\"pspReference\":\"853605802634337B\",\"response\":\"[refund-received]\"}",
+        "request": "{\"modificationAmount\":{\"value\":500,\"currency\":\"EUR\"},\"originalReference\":\"883605782632488F\",\"merchantAccount\":\"CommercetoolsGmbHDE775\"}",
+        "type": "refund"
+      }
+    }
+  ]
+}
+```
+
+### Recommendations
```
Comment (lojzatran) [inline extension/docs/Refund.md:104]: Add a note about this: https://docs.adyen.com/checkout/refund#unreferenced-refund

Comment (lojzatran) [inline extension/docs/Refund.md:104]:
```suggestion
### Additional information
```

File: extension/src/paymentHandler/payment-handler.js [hunk 5]
```diff
@@ -36,7 +38,16 @@ async function handlePayment (paymentObject) {
   return { success: true, data: handlerResponse }
 }
 
+function _isRefund (paymentObject) {
```
Comment (JudeNiroshan) [inline extension/src/paymentHandler/payment-handler.js:41]: we could move this method to utility class

File: extension/src/paymentHandler/refund-payment.handler.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
+const Promise = require('bluebird')
+const pU = require('./payment-utils')
+const { refund } = require('../service/web-component-service')
+const {
+  CTP_INTERACTION_TYPE_REFUND,
+} = require('../config/constants')
+
+async function execute (paymentObject) {
+  const refundInitTransactions = pU.listRefundTransactionsInit(paymentObject)
```
Comment (JudeNiroshan) [inline extension/src/paymentHandler/refund-payment.handler.js:9]: what if we do not find any refund transactions?

File: extension/src/paymentHandler/refund-payment.handler.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  const chargeSuccessTransaction = pU.getChargeTransactionSuccess(paymentObject)
+
+  const actions = []
```
Comment (JudeNiroshan) [inline extension/src/paymentHandler/refund-payment.handler.js:12]:
```suggestion
  const updateActions = []
```

File: extension/test/integration/refund-payment.handler.spec.js [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    const result = await ctpClient.create(ctpClient.builder.payments, paymentDraft)
+    payment = result.body
+  })
+
+  afterEach(async () => {
+    await iTSetUp.stopRunningServers()
+    await iTSetUp.cleanupCtpResources(ctpClient)
+  })
+
+  it('given a payment with "charge success transaction" ' +
```
Comment (JudeNiroshan) [inline extension/test/integration/refund-payment.handler.spec.js:55]: can we also add a test which trying to refund more the payment value?

**Review by JudeNiroshan — CHANGES_REQUESTED** [review]

File: extension/src/paymentHandler/refund-payment.handler.js (hunk 6 above)
Comment (lojzatran) [reply extension/src/paymentHandler/refund-payment.handler.js:9]: If we don't find any refund transactions, then it is not a refund and this method will not be executed.

File: extension/src/paymentHandler/refund-payment.handler.js (hunk 7 above)
Comment (lojzatran) [reply extension/src/paymentHandler/refund-payment.handler.js:12]: We name it actions in other handlers as well. It's not bad name, so I would keep it.

File: extension/test/integration/refund-payment.handler.spec.js (hunk 8 above)
Comment (lojzatran) [reply extension/test/integration/refund-payment.handler.spec.js:55]: It doesn't make sense here, there's no difference if you refund more or not. We don't check it, it's up to Adyen to check.

File: extension/src/paymentHandler/payment-handler.js (hunk 5 above)
Comment (lojzatran) [reply extension/src/paymentHandler/payment-handler.js:41]: Why? It is used only in this class.

File: extension/docs/Refund.md (hunk 4 above)
Comment (lojzatran) [reply extension/docs/Refund.md]: https://github.com/commercetools/commercetools-adyen-integration/pull/480/commits/fc5d295e2021b0c5f64b43234079d650b74610d6

**Review by JudeNiroshan — APPROVED** [review]

**Pull request merged**